File: packages/sdk-ts/src/multikey.ts
/**
 * Multikey encoding for verification methods.
 *
 * Mirrors `vouch/multikey.py`. Per the Controlled Identifiers
 * specification, public keys in DID Documents are encoded as a Multikey:
 *
 *   publicKeyMultibase = base58btc( multicodec_prefix || raw_public_key_bytes )
 *
 * The leading 'z' character indicates base58btc encoding. Cross-implementation
 * interop with the Python module is REQUIRED.
 *
 * Supported algorithms (Specification §13.5):
 *   Ed25519    multicodec prefix 0xed01  (32-byte key)
 *   ML-DSA-44   multicodec prefix 0x1207  (1312-byte key, provisional)
 */

// Multicodec prefixes as 2-byte sequences.
export const ED25519_PUB_PREFIX = new Uint8Array([0xed, 0x01]);
export const ED25519_PRIV_PREFIX = new Uint8Array([0x80, 0x26]);
export const MLDSA44_PUB_PREFIX = new Uint8Array([0x87, 0x24]);
export const MLDSA44_PRIV_PREFIX = new Uint8Array([0x88, 0x24]);

const PREFIX_TABLE: Array<[Uint8Array, string, 'public' | 'private']> = [
  [ED25519_PUB_PREFIX, 'Ed25519', 'public'],
  [ED25519_PRIV_PREFIX, 'Ed25519', 'private'],
  [MLDSA44_PUB_PREFIX, 'ML-DSA-44', 'public'],
  [MLDSA44_PRIV_PREFIX, 'ML-DSA-44', 'private'],
];

const B58_ALPHABET = '123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz';

/**
 * Encode a 32-byte Ed25519 public key as a Multikey string (z-prefixed
 * base58btc).
 */
export function encodeEd25519Public(rawKey: Uint8Array): string {
  if (rawKey.length !== 32) {
    throw new Error(`Ed25519 public key must be 32 bytes, got ${rawKey.length}`);
  }
  const buf = concat(ED25519_PUB_PREFIX, rawKey);
  return 'z' + b58encode(buf);
}

/**
 * Encode a 1312-byte ML-DSA-44 public key as a Multikey string. Used in
 * the hybrid post-quantum profile (Specification §13.2).
 */
export function encodeMLDSA44Public(rawKey: Uint8Array): string {
  if (rawKey.length !== 1312) {
    throw new Error(`ML-DSA-44 public key must be 1312 bytes, got ${rawKey.length}`);
  }
  const buf = concat(MLDSA44_PUB_PREFIX, rawKey);
  return 'z' + b58encode(buf);
}

/**
 * Decode a Multikey string into algorithm + raw key bytes.
 */
export function decode(multikey: string): { algorithm: string; rawKey: Uint8Array } {
  if (!multikey.startsWith('z')) {
    throw new Error('Multikey must use base58btc encoding (z-prefix)');
  }
  const decoded = b58decode(multikey.slice(1));
  if (decoded.length < 2) throw new Error('Multikey too short');
  const prefix = decoded.slice(0, 2);

  for (const [pat, alg, role] of PREFIX_TABLE) {
    if (prefix[0] === pat[0] && prefix[1] === pat[1]) {
      if (role === 'private') {
        // A verificationMethod must carry a PUBLIC key. Refuse a private-key
        // multicodec prefix so private material is never treated as a key.
        throw new Error('Multikey carries a private-key prefix; a public key is required');
      }
      const rawKey = decoded.slice(2);
      const expected = alg === 'Ed25519' ? 32 : 1312;
      if (rawKey.length !== expected) {
        throw new Error(`${alg} public key must be ${expected} bytes, got ${rawKey.length}`);
      }
      return { algorithm: alg, rawKey };
    }
  }
  throw new Error(
    `Unknown multicodec prefix: ${[...prefix].map(b => b.toString(16).padStart(2, '0')).join('')}`
  );
}

/**
 * Return the algorithm name encoded in a Multikey without exposing raw bytes.
 */
export function algorithmOf(multikey: string): string {
  return decode(multikey).algorithm;
}

// ---------------------------------------------------------------------------
// base58btc primitive (vendored to avoid dependencies)
// ---------------------------------------------------------------------------

export function b58encode(data: Uint8Array): string {
  if (data.length === 0) return '';

  let nZero = 0;
  for (const b of data) {
    if (b === 0) nZero++;
    else break;
  }

  // Convert bytes to a big integer
  let num = 0n;
  for (const b of data) {
    num = (num << 8n) + BigInt(b);
  }

  let encoded = '';
  while (num > 0n) {
    const rem = Number(num % 58n);
    num = num / 58n;
    encoded = B58_ALPHABET[rem] + encoded;
  }

  return '1'.repeat(nZero) + encoded;
}

export function b58decode(s: string): Uint8Array {
  if (s.length === 0) return new Uint8Array();

  let nZero = 0;
  for (const ch of s) {
    if (ch === '1') nZero++;
    else break;
  }

  let num = 0n;
  for (const ch of s) {
    const idx = B58_ALPHABET.indexOf(ch);
    if (idx < 0) throw new Error(`Invalid base58 character: ${ch}`);
    num = num * 58n + BigInt(idx);
  }

  const bytes: number[] = [];
  while (num > 0n) {
    bytes.unshift(Number(num & 0xffn));
    num = num >> 8n;
  }

  const out = new Uint8Array(nZero + bytes.length);
  for (let i = 0; i < bytes.length; i++) out[nZero + i] = bytes[i];
  return out;
}

function concat(a: Uint8Array, b: Uint8Array): Uint8Array {
  const out = new Uint8Array(a.length + b.length);
  out.set(a, 0);
  out.set(b, a.length);
  return out;
}

File: vouch/multikey.py
"""
Multikey encoding for verification methods.

Per the Controlled Identifiers specification, public keys in DID Documents
are encoded as a `Multikey` with the format:

  publicKeyMultibase = base58btc( multicodec_prefix || raw_public_key_bytes )

The leading `z` character of the multibase output indicates base58btc.

Supported algorithms (Vouch Protocol Specification §13.5):
  Ed25519    multicodec prefix 0xed01 (32-byte key)
  ML-DSA-44   multicodec prefix 0x1207 (1312-byte key, provisional)

This module provides encode/decode helpers and is dependency-free.
"""

from __future__ import annotations

from typing import Tuple


# Multicodec prefixes (varint-encoded as 2 bytes for these values).
ED25519_PUB_PREFIX = bytes([0xED, 0x01])
ED25519_PRIV_PREFIX = bytes([0x80, 0x26])  # ed25519-priv multicodec
MLDSA44_PUB_PREFIX = bytes([0x87, 0x24])  # provisional 0x1207 varint-encoded
MLDSA44_PRIV_PREFIX = bytes([0x88, 0x24])  # provisional

# Reverse lookup
_PREFIX_TO_ALG = {
    ED25519_PUB_PREFIX: ("Ed25519", "public"),
    ED25519_PRIV_PREFIX: ("Ed25519", "private"),
    MLDSA44_PUB_PREFIX: ("ML-DSA-44", "public"),
    MLDSA44_PRIV_PREFIX: ("ML-DSA-44", "private"),
}

# base58btc alphabet (Bitcoin's standard alphabet)
_B58_ALPHABET = b"123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz"
_B58_INDEX = {c: i for i, c in enumerate(_B58_ALPHABET)}


def encode_ed25519_public(raw_key: bytes) -> str:
    """Encode a 32-byte Ed25519 public key as a Multikey string (z-prefixed base58btc)."""
    if len(raw_key) != 32:
        raise ValueError(f"Ed25519 public key must be 32 bytes, got {len(raw_key)}")
    return "z" + _b58encode(ED25519_PUB_PREFIX + raw_key)


def encode_mldsa44_public(raw_key: bytes) -> str:
    """Encode an ML-DSA-44 public key (1312 bytes) as a Multikey string.

    Used in the hybrid post-quantum profile (Specification §13.2). The
    DID Document publishes this alongside an Ed25519 Multikey under a
    parallel `verificationMethod` slot.
    """
    if len(raw_key) != 1312:
        raise ValueError(f"ML-DSA-44 public key must be 1312 bytes, got {len(raw_key)}")
    return "z" + _b58encode(MLDSA44_PUB_PREFIX + raw_key)


def decode(multikey: str) -> Tuple[str, bytes]:
    """
    Decode a Multikey string. Returns (algorithm_name, raw_key_bytes).

    Raises ValueError if the format is unrecognized or the key is malformed.
    """
    if not multikey.startswith("z"):
        raise ValueError("Multikey must use base58btc encoding (z-prefix)")
    decoded = _b58decode(multikey[1:])
    if len(decoded) < 2:
        raise ValueError("Multikey too short")
    prefix = decoded[:2]
    if prefix not in _PREFIX_TO_ALG:
        raise ValueError(f"Unknown multicodec prefix: {prefix.hex()}")
    alg, kind = _PREFIX_TO_ALG[prefix]
    if kind == "private":
        # A verificationMethod must carry a PUBLIC key. Refuse a private-key
        # multicodec prefix so private material is never treated as a key.
        raise ValueError("Multikey carries a private-key prefix; a public key is required")
    raw = decoded[2:]
    expected = 32 if alg == "Ed25519" else 1312
    if len(raw) != expected:
        raise ValueError(f"{alg} public key must be {expected} bytes, got {len(raw)}")
    return alg, raw


def algorithm_of(multikey: str) -> str:
    """Return the algorithm name encoded in `multikey` without exposing key bytes."""
    alg, _ = decode(multikey)
    return alg


# ---- base58btc primitive (vendored to avoid a dependency) ----


def _b58encode(data: bytes) -> str:
    if not data:
        return ""
    n_zero = 0
    for byte in data:
        if byte == 0:
            n_zero += 1
        else:
            break
    num = int.from_bytes(data, "big")
    encoded = bytearray()
    while num > 0:
        num, rem = divmod(num, 58)
        encoded.append(_B58_ALPHABET[rem])
    encoded.extend(b"1" * n_zero)
    encoded.reverse()
    return encoded.decode("ascii")


def _b58decode(s: str) -> bytes:
    if not s:
        return b""
    n_zero = 0
    for ch in s:
        if ch == "1":
            n_zero += 1
        else:
            break
    num = 0
    for ch in s:
        idx = _B58_INDEX.get(ord(ch))
        if idx is None:
            raise ValueError(f"Invalid base58 character: {ch!r}")
        num = num * 58 + idx
    body = num.to_bytes((num.bit_length() + 7) // 8, "big") if num else b""
    return b"\x00" * n_zero + body
